### ISSUE TITLE
livepush: CLI exits unexpectedly (no error) with Node.js v14

### ISSUE THREAD

**Issue by pdcastro** [balena-io/balena-cli#2165]
As originally reported in a [flowdock thread](https://www.flowdock.com/app/rulemotion/resin-starters/threads/k2KO6r6D4ep5CyKC_Sf-DaU01Po) (private):

> livepushing my BoB:
```
[Debug]   Found project types:
[Debug]       admin: External image
[Debug]       builder: External image
[Debug]       db: External image
[Debug]       haproxy: External image
[Debug]       s3: External image
[Debug]       redis: External image
[Debug]       balena-mdns-publisher: External image
[Debug]   Probing remote daemon for cache images
[Debug]   Using 586 on-device images for cache...
[Debug]   Starting builds...

~/balena/balena-on-balena master* 1m 5s
```
> basically my push just craps out with no error/log. there's also nothing supervisor-side. this is the latest version 12.38.3 and the latest supervisor 12.3.0. i've also commented out nearly all of my services, as you can tell from the logs above

This issue only applies to the advanced `npm` installation of balena CLI, where the system installation of Node.js is used. It does not apply to the standalone zip package (all platforms) or the GUI installers (Windows and Mac), as these options currently use a built-in Node.js v12.

Possibly related to #1847 or #1878 (maybe not).

**Known workarounds:**
* Use the standalone zip package (all platforms)
* Use the GUI installer (Windows or macOS)
* If using the `npm` installation, downgrade the Node.js to v10 or v12, instead of v14.

**Comment by LucianBuzzo** [comment]
@pdcastro Do you recommend switching to node v12 for a better `balena push local` experience?

**Comment by LucianBuzzo** [comment]
I'm developing on an intel NUC (Os: balenaOS 2.68.1+rev1, super 12.3.0) and my main workstation is windows 10 running WSL 1.
I was having a lot of issues with livepush aborting without an error message or segfaulting. I wasn't able to ctrl+c the build and restart it without it failing and the only way to get it working again would be to disable then re-enable local mode. 
I switched from node 14 to node 12 and it is now working flawlessly 👍

**Comment by StefKors** [comment]
@pdcastro is this the same as https://github.com/balena-io/balena-supervisor/issues/1568 ?

**Comment by pdcastro** [comment]
> Do you recommend switching to node v12 for a better `balena push local` experience?

Yes. The CLI does not yet fully support Node v14. The CLI's GUI installers (Windows / Mac) and standalone zip packages (all platforms) ship with Node v12. The "advanced" `npm` installation method [documents](https://github.com/balena-io/balena-cli/blob/master/INSTALL-ADVANCED.md#npm-installation) that Node v14 is not yet supported. Starting with CLI v12.38.4 (2021-01-21), we have also further restricted the required versions through the [engines section of package.json](https://github.com/balena-io/balena-cli/blob/v12.38.8/package.json#L87), which now reads `"node": ">=10.20.0 <13.0.0"`. As a result, since that CLI release, running the CLI with Node v14 produces a warning message:
```
$ nvm use 14
Now using node v14.15.4 (npm v6.14.10)

$ balena apps
------------------------------------------------------------------------------
Warning: Node version "v14.15.4" does not match required versions ">=10.20.0 <13.0.0".
This may cause unexpected behavior. To upgrade Node, visit:
https://nodejs.org/en/download/
------------------------------------------------------------------------------
Login required: use the “balena login” command to log in.
```
(Btw, that warning message is the default message produced by Node.js, it's not in the CLI code. If it was in CLI code, I would have reworded and probably removed that nodejs.org URL.)

cc: @mbalamat

**Comment by pdcastro** [comment]
Some findings were discussed in (restricted access) https://www.flowdock.com/app/rulemotion/i-cli/threads/m_GgSMpEVxho3gktOKD6_xBB9GV

**Comment by taromurao** [comment]
I was able to `balena push` with Node 14. Way to reproduce

1. Install Node 14 with nvm

```
❯ nvm list
     system  
   v12.22.9 lts/erbium 
 ▶ v14.18.3 lts/fermium 
   v16.13.2 lts/gallium 
    v17.3.1 latest 
 ```
 
2. Create a fleet with RPi 4
3. Git clone https://github.com/balena-io/balena-cli-advanced-masterclass
4. Start up the device and make it `local-mode`
5. Inside the cloned repository, run `balena push <device UUID>`

Can you also try @pdcastro ? Probably difference in Node minor version makes a difference? My dev machine has Manjaro with Linux kernel 5.16, but OS should not affect???

**Comment by pdcastro** [comment]
@taromurao, thanks for testing it. 👍  This issue has only ever been consistently reproduced with "heavy" projects, such as the BoB repo ( https://github.com/balena-io/balena-on-balena/ ) and I believe also the Jellyfish project.

"Simple projects" like your example have worked with Node.js v14 for a very long time.

Presumably, customer projects that are as "heavy" as BoB or Jellyfish would also hit this issue.

"Heavy" may be related to CPU, RAM, network, disk activity, number of services, a combination of these or something else -- we are not really sure.

Try pushing the BoB repo if you can, for example to the point that you can open the web dashboard running on BoB. I have not done it in a while, and since then lots of changes were made to both the CLI and to the BoB project. I believe the latest instructions for setting up BoB are at https://github.com/balena-io/balena-on-balena/blob/master/docs/development.md

It is possible that this issue is no longer reproducible with BoB and Jellyfish, because of all the changes to CLI dependencies in the past year. If it is no longer reproducible with BoB and Jellyfish, then we could declare it fixed and proceed with testing Node.js v16 and later. :-)

**Comment by taromurao** [comment]
@pdcastro I just tried to llivpush BoB, and it indeed crapped out without any error message. It seems to have happened when download of all container images were just finished. I have not yet found anything particular.

**Comment by taromurao** [comment]
@pdcastro Today I tested Node 14 again because recently there has been an update of the dependency "ssh2".  Pushing Jellyfish went through. Bob failed with

```
admin: (HTTP code 404) unexpected - pull access denied for balena/balena-admin, repository does not exist or may require 'docker login': denied: requested access to the resource is denied 
	api: (HTTP code 404) unexpected - pull access denied for balena/balena-api, repository does not exist or may require 'docker login': denied: requested access to the resource is denied 
	builder: (HTTP code 404) unexpected - pull access denied for balena/balena-builder, repository does not exist or may require 'docker login': denied: requested access to the resource is denied 
	delta: (HTTP code 404) unexpected - pull access denied for balena/balena-delta, repository does not exist or may require 'docker login': denied: requested access to the resource is denied 
	devices: (HTTP code 404) unexpected - pull access denied for balena/balena-proxy, repository does not exist or may require 'docker login': denied: requested access to the resource is denied 
	git: (HTTP code 404) unexpected - pull access denied for balena/balena-git, repository does not exist or may require 'docker login': denied: requested access to the resource is denied 
	img: (HTTP code 404) unexpected - pull access denied for balena/balena-img, repository does not exist or may require 'docker login': denied: requested access to the resource is denied 
	ui: (HTTP code 404) unexpected - pull access denied for balena/balena-ui, repository does not exist or may require 'docker login': denied: requested access to the resource is denied 
	sentry: (HTTP code 404) unexpected - pull access denied for balena/balena-sentry, repository does not exist or may require 'docker login': denied: requested access to the resource is denied 
	monitor: (HTTP code 404) unexpected - pull access denied for balena/balena-monitor, repository does not exist or may require 'docker login': denied: requested access to the resource is denied 
	alertmanager: (HTTP code 404) unexpected - pull access denied for balena/balena-alertmanager, repository does not exist or may require 'docker login': denied: requested access to the resource is denied 
	test-bob: pull access denied for balena/e2e-browser-tests, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
	autohat: pull access denied for balena/autohat, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```

**Comment by pipex** [comment]
I just remembered clearly when I saw a similar behavior. I was trying to build an image programmatically (loading the local files into a tar stream and then sending those files to the docker engine), but I neglected to remove files referenced in dockerignore, which meant I was adding all the files under `node_modules` to a tar stream living in memory. In that case the process would just terminate without reporting any specific errors, such as the issue describes.
